0002-crypto-conditions/0002-crypto-conditions.md

```diff
@@ -717,102 +638,151 @@ This section to be expanded in a later draft.  <!-- TODO --> For now, see the te
+        3975D78A 98B0821E E55CA0F8 6305D425 29E10EB0 15CEFD40 2FB59B2A BB8DEEE5
+        2A6F2447 D2284603 D219CD4E 8CF9CFFD D5498889 C3780B59 DD6A57EF 7D732620
+        'H
     }
 
     exampleEd25519FulfillmentPayload Ed25519FulfillmentPayload ::=
     {
-    publicKey '
-    EC172B93 AD5E563B F4932C70 E1245034 C35467EF 2EFD4D64 EBF81968 3467E2BF
-    'H,
-    signature '
-    B62291FA D9432F8F 298B9C4A 4895DBE2 93F6FFDA 1A68DADF 0CCDEF5F 47A0C721
-    2A5FEA3C DA97A3F4 C03EA9F2 E8AC1CEC 86A51D45 2127ABDB A09D1B6F 331C070A
-    'H
+        publicKey '
+        EC172B93 AD5E563B F4932C70 E1245034 C35467EF 2EFD4D64 EBF81968 3467E2BF
+        'H,
+        signature '
+        B62291FA D9432F8F 298B9C4A 4895DBE2 93F6FFDA 1A68DADF 0CCDEF5F 47A0C721
+        2A5FEA3C DA97A3F4 C03EA9F2 E8AC1CEC 86A51D45 2127ABDB A09D1B6F 331C070A
+        'H
     }
 
     END
 
-
 # IANA Considerations {#appendix-e}
 
 ## Crypto-Condition Type Registry {#crypto-conditions-type-registry}
 
 The following initial entries should be added to the Crypto-Condition Type registry to be created and maintained at (the suggested URI)
 <http://www.iana.org/assignments/crypto-condition-types>:
 
-The following feature suite bits are registered:
+The following types are registered:
 
-| Type Bit | Exp. | Hex  | Feature         |
-|----------|------|------|-----------------|
-| 1        | 2^0  | 0x01 | SHA-256         |
-| 10       | 2^1  | 0x02 | PREIMAGE        |
-| 100      | 2^2  | 0x04 | PREFIX          |
-| 1000     | 2^3  | 0x08 | THRESHOLD       |
-| 10000    | 2^4  | 0x10 | RSA             |
-| 100000   | 2^5  | 0x20 | ED25519         |
-{: #crypto-condition-feature-suites title="Crypto-Condition Feature Suites"}
+| Type ID | Type Name         |
+|---------|-------------------|
+| 0       | PREIMAGE-SHA-256  |
+| 1       | PREFIX-SHA-256    |
+| 2       | THRESHOLD-SHA-256 |
+| 3       | RSA-SHA-256       |
+| 4       | ED25519           |
+{: #crypto-condition-types-table title="Crypto-Condition Types"}
 
-The following types are registered:
+# String Encoding {#appendix-f}
+
+Implementations MAY support one or both string encoding formats which encode conditions and fulfillments as either URIs or JSON objects. The binary encoding is considered the canonical encoding.
+
+The following string encoding types are defined:
+
+BASE10
+: Variable-length integer encoded as a base-10 (decimal) number. Implementations MUST reject encoded values that are too large for them to parse. Implementations MUST be tested for overflows.
+
+BASE16
+: Variable-length integer encoded as a base-16 (hexadecimal) number. Implementations MUST reject encoded values that are too large for them to parse. Implementations MUST be tested for overflows. Encodings may have an odd number of characters as the encoding excludes leading zeros.
+
+BASE64URL
+: Base64-URL encoding. See [RFC4648](#RFC4648), Section 5.
+
+## Condition URI Format {#string-condition-format}
```

Comment:
@nbougalis Your argument is spot on when it comes to the fulfillment. The fulfillment is a huge OER encoded blob. And the only things you'll want to know about it is its structure and whether it's valid. And to know either you need to be able to fully parse it anyway. There is very, very little value in knowing just the top level type of the fulfillment and nothing else. So having a top level string encoding makes no sense.

For the condition, it's kinda the opposite. Conditions have no structure past the top level. They consist only of a hash, a type, a max fulfillment length and maybe a subtypes bitmap. None of these things require OER to parse. I could easily see people wanting to build simple filters that can check whether a condition type is supported. It's also nice to be able to see the type just from looking at the condition, e.g. `cc:preimage-sha256:...`

The point is: I think it's very important to answer this question for the condition and the fulfillment separately because the design goals in each case are very different.

For the condition, I like the idea of sticking with something similar to the current format, like `cc:preimage-sha256:[hash]:1024` and `cc:threshold-sha256:[hash]:[subtypes bitmap]:15300`. For the fulfillment, I like the suggestion of switching to `"cf:" BASE64URL(TO_BLOB(F))`

There is still the issue that conditions also appear in threshold and prefix fulfillments as "subconditions". So we either need to have a binary encoding in addition to the string encoding (awkward) or we have to use the string encoding inside of the binary encoding (awkward).

We may be able to reduce the awkwardness of the latter option. Instead of having the fingerprint contents of a threshold condition include the condition as a string (e.g. `"cc:preimage-sha256:[hash]:1024"`), they could include a hash of the condition string instead (e.g. `0x13abd...`), then it would just be a Uint256 in the OER format, which saves a few bytes, is just as secure and solves the bad taste of putting string format inside of a binary format.